### PR TITLE
feat:(ignoreValues): Adding flag for single-stage validation.

### LINKED ISSUE #17: Feature Request: Support for a single-step verification
### Is your feature request related to a problem?

This policy as-is requires that default max values be provided for either/both CPU and Memory.

I would like to be able to accept/reject based on the existence of CPU/Memory requests/limits at all, and don't care what they are set to. I only care that they are set.

In Rancher™, we specify namespace / project resource limits and do not need to enforce them individually.

### Solution you'd like

I'd like to be able to provide the following YAML for the policy settings, which would respectively be a two-step verification, or a single.

```yaml
# optional
memory:
  defaultRequest: "100M"
  defaultLimit: "500M"
  maxLimit: "4G"
# optional - Just needs to exist.
cpu:
  defaultRequest: "0M"
  defaultLimit: "0M"
  maxLimit: "0G"
```

### Alternatives you've considered

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by TechDufus**
This PR introduces a new configuration setting for both `cpu` and `memory` settings: `ignoreValues`

## Description
The purpose of this addition is to support enforcing that `cpu` and `memory` configuration is set, without actually caring what it is set to.

In my example (why I wanted this feature), is we enforce `cpu` and `memory` limits via [Rancher](https://www.rancher.com/) at the `project` and `namespace` level, and as a best practice, simply want to enforce that our dev teams are configuring these themselves as a best practice but we do not care what they set them to.

Fix #17 

## Test

To test this pull request, you can run the following commands:

```shell
cd <to_package_directory>
make e2e-tests
```

### Potential improvement

This allows this policy to be used in many more scenarios without taking away from it's original / core purpose and functionality.

### CODE REVIEW

File: settings_test.go [hunk 1]
```diff
@@ -177,4 +180,72 @@ func TestNewSettingsPartialFieldsOnlyFromValidationReq(t *testing.T) {
...
+		if err != nil {
+			t.Fatalf("Unexpected error %+v", err)
+		}
+		if !settings.Cpu.IgnoreValues {
+			t.Errorf("ignoreValues should be true")
+		}
+		if !settings.Memory.IgnoreValues {
+			t.Errorf("ignoreValues should be true")
+		}
+	})
```
Comment (jvanz) [inline settings_test.go:250]: Thanks for adding these tests! But I think we are starting to have too much duplicate code in the tests. Can you try to move the validation code into a function and call it passing the `Cpu` or `Memory` and the expected values. Something line this:

```
if err := checkSettingsValues(settings.Cpu, expectedMaxLimit, expectedDefaultLimit, expectedDefaultRequest, expectedIgnoreValues); err != nil {
    t.Errorf("invalid Cpu settings: %v", err)

}
```

File: test_data/pod_without_resources.json [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+          "image": "image:latest",
+          "resources": {
+            "limits": {
+              "cpu": "2m",
+              "memory": "2Gi"
+            },
+            "requests": {
+              "cpu": "2m",
+              "memory": "2Gi"
+            }
```
Comment (jvanz) [inline test_data/pod_without_resources.json:61]: The file name does not match what is in the file. Should not the `resource` field be empty?

File: validate.go
Comment (jvanz) [inline validate.go:1]: In this file you are ignoring the values only in the limits validation. However, in the docs you say that the settings is used in limit and requests. I think you need to update the `validateAndAdjustContainerResourceRequests` and `adjustResourceRequest` functions to check if the request resources are defined. Therefore, if the `ignoreValues` is `true` and the requested resources are missing, return an error. If `ignoreValues` is `false`, do what is done before (mutate the request)

**Review by jvanz — CHANGES_REQUESTED** [review]
Thanks for doing this! The change looks pretty good. I just have improvements request.

File: test_data/pod_without_resources.json (hunk 2 above)
Comment (TechDufus) [reply test_data/pod_without_resources.json]: Yes it should, I just made [this change](https://github.com/kubewarden/container-resources-policy/pull/18/commits/8c968566a702748292f061e945df3e86bfc053c1). I'm not able to run the bats tests since I can't build the policy [at the moment](https://github.com/kubewarden/container-resources-policy/issues/17#issuecomment-1966981012).

File: settings_test.go (hunk 1 above)
Comment (TechDufus) [reply settings_test.go:250]: Is your goal to replace most setting test logic with this `checkSettingsValues` logic? Some tests are checking with `.MustParse()` and most are just comparing strings.

Comment (TechDufus) [reply settings_test.go:193]: This should be taken care of with https://github.com/kubewarden/container-resources-policy/pull/18/commits/9bfcf0f8bb6e9f20d1e52e97317a4813e9408944

File: validate.go
Comment (TechDufus) [reply validate.go:1]: This should be resolved here https://github.com/kubewarden/container-resources-policy/pull/18/commits/b369934bd5b63f99ad815e0dcdc2f3bde77b9e2a. Thanks for catching this.

File: test_data/pod_without_resources.json (hunk 2 above)
Comment (TechDufus) [reply test_data/pod_without_resources.json]: This is resolved here https://github.com/kubewarden/container-resources-policy/pull/18/commits/8c968566a702748292f061e945df3e86bfc053c1.

File: validate.go [hunk 3]
```diff
@@ -29,10 +29,10 @@ func adjustResourceRequest(container *corev1.Container, resourceName string, res
 // Return `true` when the container has been mutated
 func validateAndAdjustContainerResourceRequests(container *corev1.Container, settings *Settings) bool {
 	mutated := false
-	if settings.Memory != nil {
+	if !settings.Memory.IgnoreValues && settings.Memory != nil {
 		mutated = adjustResourceRequest(container, "memory", settings.Memory)
 	}
-	if settings.Cpu != nil {
+	if !settings.Memory.IgnoreValues && settings.Cpu != nil {
```
Comment (flavio) [inline validate.go:35]: Wrong, it should be `settings.Cpu.IgnoreValues`, before fixing it, can you reproduce the issue with a unit test?

File: validate.go [hunk 4]
```diff
@@ -41,6 +41,9 @@ func validateAndAdjustContainerResourceRequests(container *corev1.Container, set
 func validateAndAdjustContainerResourceLimit(container *corev1.Container, resourceName string, resourceConfig *ResourceConfiguration) (bool, error) {
```
Comment (flavio) [inline validate.go:41]: Unrelated with this PR, could you add some inline docs about what this function is doing? It took me a while to understand it's also mutating the `container` object given a input and understand the `bool` value returned is not about something being valid/invalid; but rather the `container` being mutated or not

Comment (flavio) [reply validate.go:41]: Also, can you update the inline documentation of `validateAndAdjustContainerResourceLimits` to reflect the behavior change introduced by this PR?

**Review by flavio — CHANGES_REQUESTED** [review]
Thanks, this looks great. I left some comments

File: settings_test.go (hunk 1 above)
Comment (jvanz) [reply settings_test.go:193]: That's better! Thanks!

File: validate.go (hunk 3 above)
Comment (jvanz) [reply validate.go:35]: Furthermore, I think you need to invert the && values. If the `settings.Cpu` is nil, the program will panic.

File: validate.go [hunk 5]
```diff
@@ -29,10 +29,10 @@ func adjustResourceRequest(container *corev1.Container, resourceName string, res
 // Return `true` when the container has been mutated
 func validateAndAdjustContainerResourceRequests(container *corev1.Container, settings *Settings) bool {
 	mutated := false
-	if settings.Memory != nil {
+	if !settings.Memory.IgnoreValues && settings.Memory != nil {
```
Comment (jvanz) [inline validate.go:32]: I don't think this is covering what is being said in this README section:

```
If you would only like to enforce that requests and limits are set (ie, you do not care
what they are set to, just that they have been set), you can set `ignoreValues` to `true`.
This will skip the enforcement of specific values and only enforce that requests and
limits are set. Here is an example of how to configure this:
```

For example, consider this unit test:

```go
		{"container missing memory requests values and ignore memory values", corev1.Container{
			Image: "image1:latest",
			Resources: &corev1.ResourceRequirements{
				Limits: map[string]*apimachinery_pkg_api_resource.Quantity{
					"cpu": &oneCoreCpuQuantity,
					"memory": &oneGiMemoryQuantity,
				},
				Requests: map[string]*apimachinery_pkg_api_resource.Quantity{
					"cpu": &oneCoreCpuQuantity,
				},
			},
		}, Settings{
			Cpu: &ResourceConfiguration{
				DefaultLimit:   oneCore,
				DefaultRequest: oneCore,
				MaxLimit:       oneCore,
			},
			Memory: &ResourceConfiguration{
				IgnoreValues: true,
			},
		}, &corev1.ResourceRequirements{
			Limits: map[string]*apimachinery_pkg_api_resource.Quantity{
				"cpu": &oneCoreCpuQuantity,
				"memory": &oneGiMemoryQuantity,
			},
			Requests: map[string]*apimachinery_pkg_api_resource.Quantity{
				"cpu": &oneCoreCpuQuantity,
			},
		}, false, "missing memory resource request"},
// this can be added in the unit tests file. ;-)
```

The policy will not return an error saing: "hey! You need to add the memory in the requested resources".

**Review by jvanz — CHANGES_REQUESTED** [review]

File: validate.go (hunk 3 above)
Comment (TechDufus) [reply validate.go]: I've done a slight refoctor that resolves this issue and removes this condition on this IF.

File: validate.go [hunk 6]
```diff
@@ -40,7 +94,13 @@ func validateAndAdjustContainerResourceRequests(container *corev1.Container, set
 
 func validateAndAdjustContainerResourceLimit(container *corev1.Container, resourceName string, resourceConfig *ResourceConfiguration) (bool, error) {
 	resourceStr, found := container.Resources.Limits[resourceName]
+	if resourceConfig.IgnoreValues && !found {
+		return false, fmt.Errorf("container does not have a %s limit", resourceName)
+	}
 	if !found || resourceStr == nil || len(strings.TrimSpace(string(*resourceStr))) == 0 {
+		if resourceConfig.IgnoreValues {
+			return false, fmt.Errorf("missing %s limit", resourceName)
+		}
```
Comment (jvanz) [inline validate.go:103]: Now that you have other functions that is checking the resource fiel before calling this function, you do not need to check the `IgnoreValues` again. You can remove this if or remove the new functions.

File: validate.go [hunk 7]
```diff
@@ -40,7 +94,13 @@ func validateAndAdjustContainerResourceRequests(container *corev1.Container, set
 
 func validateAndAdjustContainerResourceLimit(container *corev1.Container, resourceName string, resourceConfig *ResourceConfiguration) (bool, error) {
 	resourceStr, found := container.Resources.Limits[resourceName]
+	if resourceConfig.IgnoreValues && !found {
+		return false, fmt.Errorf("container does not have a %s limit", resourceName)
+	}
```
Comment (jvanz) [inline validate.go:99]: Now that you have other functions that is checking the resource fiel before calling this function, you do not need to check the `IgnoreValues` again. You can remove this if or remove the new functions.

File: validate.go [hunk 8]
```diff
@@ -24,6 +24,60 @@ func adjustResourceRequest(container *corev1.Container, resourceName string, res
...
+		return fmt.Errorf("container does not have any resource limits")
+	}
+
+	cpuStr, found := container.Resources.Limits["cpu"]
+	if settings.Cpu.IgnoreValues && (!found || cpuStr == nil || len(strings.TrimSpace(string(*cpuStr))) == 0) {
+		return fmt.Errorf("container does not have a cpu limit")
+	}
+
+	memoryStr, found := container.Resources.Limits["memory"]
+	if settings.Memory.IgnoreValues && (!found || memoryStr == nil || len(strings.TrimSpace(string(*memoryStr))) == 0) {
```
Comment (jvanz) [inline validate.go:38]: We use the same piece of code to check if there is something in the resource at 3 different places. Let's extract that into a function. Like this:

```suggestion
	if settings.Memory.IgnoreValues && missingResourceQuantity(container.Resource.Limits, "memory"){
```
The function code could be something like:

```go
func missingResourceQuantity(resources map[string]*apimachinery_pkg_api_resource.Quantity, resourceName string) bool{

resourceStr, found := resources[resourceName]
return !found || resourceStr == nil || len(strings.TrimSpace(string(*resourceStr))) == 0

}
```

Call the suggested function in the `adjustResourceRequest` function as well.

File: validate.go [hunk 9]
```diff
@@ -63,6 +123,9 @@ func validateAndAdjustContainerResourceLimit(container *corev1.Container, resour
 // Return `true` when the container has been mutated.
 func validateAndAdjustContainerResourceLimits(container *corev1.Container, settings *Settings) (bool, error) {
 	mutated := false
+	if settings.Memory.IgnoreValues {
+		return false, nil
+	}
```
Comment (jvanz) [inline validate.go:128]: Now that you have other functions that is checking the resource fiel before calling this function, you do not need to check the IgnoreValues again. You can remove this if or remove the new functions.

File: validate.go [hunk 10]
```diff
@@ -71,6 +134,9 @@ func validateAndAdjustContainerResourceLimits(container *corev1.Container, setti
 		}
 	}
 
+	if settings.Cpu.IgnoreValues {
+		return false, nil
+	}
```
Comment (jvanz) [inline validate.go:139]: Now that you have other functions that is checking the resource fiel before calling this function, you do not need to check the IgnoreValues again. You can remove this if or remove the new functions.

**Review by jvanz — CHANGES_REQUESTED** [review]
Cool! Now I think the policy is covering all the cases. But I think we can do some improvements in the code. Please, let me know what do you think!

 Thank you for the pacience in the review process! :)

File: validate.go (hunk 6 above)
Comment (TechDufus) [reply validate.go:103]: Nice catch... This WAS an issue until I moved around some unit tests. Thanks for catching this!

File: validate.go (hunk 8 above)
Comment (TechDufus) [reply validate.go]: This is taken care of here. https://github.com/kubewarden/container-resources-policy/pull/18/commits/f68a538ec29feb1afe0ffdd995e96f67f4e2f4f2

File: validate.go [hunk 11]
```diff
@@ -24,6 +28,58 @@ func adjustResourceRequest(container *corev1.Container, resourceName string, res
 	return false
 }
 
+func validateContainerResourceLimits(container *corev1.Container, settings *Settings) error {
+	if container.Resources.Limits == nil && settings.Cpu.IgnoreValues && settings.Memory.IgnoreValues {
+		return fmt.Errorf("container does not have any resource limits")
+	}
+
+	if settings.Cpu.IgnoreValues && (missingResourceQuantity(container.Resources.Limits, "cpu")) {
```
Comment (jvanz) [inline validate.go:36]:
```suggestion
	if settings.Cpu.IgnoreValues && missingResourceQuantity(container.Resources.Limits, "cpu") {
```

File: validate.go [hunk 12]
```diff
@@ -24,6 +28,58 @@ func adjustResourceRequest(container *corev1.Container, resourceName string, res
...
+		return fmt.Errorf("container does not have a cpu limit")
+	}
+
+	if settings.Memory.IgnoreValues && (missingResourceQuantity(container.Resources.Limits, "memory")) {
```
Comment (jvanz) [inline validate.go:40]:
```suggestion
	if settings.Memory.IgnoreValues && missingResourceQuantity(container.Resources.Limits, "memory") {
```

**Review by jvanz — CHANGES_REQUESTED** [review]
Thanks for the changes! I noticed that removing those if statements that I've asked in a previous review, introduce a bug and make the tests to fails. Hence, I've revert the commit where you do that change. Furthermore, I've added another little e2e test to cover another setting validation case. 

I would like to fix the CI before merging this and just apply the last little request changes. Besides that, LGTM

**Comment by TechDufus** [comment]
Glad to hear it! I'm only able to run the unit tests locally since I can't build and run the bats tests for some reason... Still looking into that. THANK YOU for helping me test, and especially with your patience with my code additions! 💖

**Review by jvanz — APPROVED** [review]

**Comment by jvanz** [comment]
@flavio can you take another look as well?

File: validate.go [hunk 13]
```diff
@@ -39,14 +95,14 @@ func validateAndAdjustContainerResourceRequests(container *corev1.Container, set
 }
 
 func validateAndAdjustContainerResourceLimit(container *corev1.Container, resourceName string, resourceConfig *ResourceConfiguration) (bool, error) {
```
Comment (viccuad) [inline validate.go:97]:
```suggestion
// validateAndAdjustContainerResourceLimit validates the container against the passed resourceConfig.
// If not passing validation, it mutates the container and returns true. If passing validation without the
// need of mutation, it returns false.
func validateAndAdjustContainerResourceLimit(container *corev1.Container, resourceName string, resourceConfig *ResourceConfiguration) (bool, error) {
```

**Review by viccuad — APPROVED** [review]
LGTM! Thanks! Added the function godocs that Flavio requested.

**Review by flavio — APPROVED** [review]
@TechDufus:  thanks for this contribution and for all the changes done during the review process

**Comment by TechDufus** [comment]
> @TechDufus: thanks for this contribution and for all the changes done during the review process

Thanks for letting me work on it! You guys rock.
